### DISCUSSION TITLE
How to use WS-Security with Nonce and Created properties.

### DISCUSSION THREAD

**Discussion by BelvedereHenrique** [CoreWCF/CoreWCF#1400] [Q&A]
Hello,

I'm new to SOAP WS-Security and I'm currently trying to integrate an ASPNET Core WebApi with a SOAP WS.
The SOAP WS developers provided me some documentation regarding WS-Security standards and I'm not sure about some points.

Here's what their docs say the header should be at the end:

```
<soapenv:Header
	xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
	<wsse:Security
		xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd"
		xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd"
		xmlns:wsa="http://www.w3.org/2005/08/addressing">
		<wsse:UsernameTokenwsu:Id="{username}">
			<wsse:Username>MyUsername</wsse:Username>
			<wsse:PasswordType="wsse:PasswordDigest">{computedPassword}</wsse:Password>
			<wsse:NonceEncodingType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#Base64Binary">{base64}</wsse:Nonce>
			<wsu:Created>{yyyy -MM - ddTHH:mm:ssZ}</wsu:Created>
		</wsse:UsernameToken>
		<ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
			<ds:SignedInfo>
				<ds:CanonicalizationMethodAlgorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				<ds:SignatureMethodAlgorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/>
				<ds:ReferenceURI="{username}">
					<ds:DigestMethodAlgorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
					<ds:DigestValue>{digestValue}</ds:DigestValue>
				</ds:Reference>
			</ds:SignedInfo>
                    <ds:SignatureValue>{signatureValue}</ds:SignatureValue>
		</ds:Signature>
	</wsse:Security>
</soapenv:Header>

```

Their WS requires a Security header on every request envelope containing: UsernameToken (Username, Password,Nonce,Created) and ds:Signature.

I read on multiple posts that WCF does not provide the `Nonce` and `Created` properties in the UsernameToken header, but also read it actually does and should be available since it's built-in. If it does not add them automatically, how do I add these two properties manually? I suppose I'll also have to manually add the computed the password , which is composed of `Base64SHA1(nonceValue+ createdDateValue + secret))`
I suppose that if `Nonce` and `Created` are automatically added, `PasswordDigest` will also be added automatically.

Is the `WSHttbinding` with `SecurityMode.TransportWithMessageCredential` the proper binding and configuration to be used in this case?
Does the ds:Signature header requires me to provide a certificate on my client? Should I use `ClientCertificate` or `ServiceCertificate`? Or both? (commented code below)

I've imported their WSDL files using `dotnet svcutil` into my project and here's my current code where I'm trying to do all these things:
```c#
            var binding = new WSHttpBinding(SecurityMode.TransportWithMessageCredential)
            {
                MessageEncoding = WSMessageEncoding.Mtom
            };
            binding.Security.Message.ClientCredentialType = MessageCredentialType.UserName;
            var address =
                new EndpointAddress("https://exampleaddress.com/WSendpoint");

            var client= new WsClient(binding, address);

            client.ClientCredentials.UserName.UserName = "TestUsername";
            client.ClientCredentials.UserName.Password = "TestPassword";


            // var certificate = new X509Certificate2("path_certificate.pfx", "certificate_password");
            // _synchronizationClient.ClientCredentials.ClientCertificate.Certificate = certificate;
            //
            // var cert = new X509ServiceCertificateAuthentication();
            // _synchronizationClient.ClientCredentials.ServiceCertificate.SslCertificateAuthentication = cert;
            
           var result = await client.DoSomething("123");
 
```

Any help with how to properly configure this would be highly appreciated.

Thank you!

**Comment by mconnew** [comment]
WCF doesn't support it because the authentication is by default backed by Windows accounts. The only way to validate a Windows account with UserName is for the client to provide the password. We don't support that scenario in CoreWCF and we could add this support without a lot of work. There's a few ways it could look, but one way is to have a type derived from UserNamePasswordValidator that you use when using PasswordDigest. There would be a new ValidateAsync method which passes the UserName and you have to return the password from the method. Then the validator would do the work of recalculating the digest to validate it, which would include handling of the nonce.  
It's not a lot of work if you have context on the code. If you want to try tackling it and submitting a PR, I can help point you to the pieces which need to be modified. Basically the moving pieces are:
* [UserNamePasswordTokenEntry](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/Security/WSSecurityJan2004.cs#L408) needs to be updated to handle the two missing fields - currently has code to skip them
* [UserNameSecurityToken](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/UserNameSecurityToken.cs) needs the two missing fields added
* [UserNameSecurityTokenHandler](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/UserNameSecurityTokenHandler.cs) needs to be updated to handle the two extra fields - currently has code to skip them
* UserNameDigestSecurityTokenAuthenticator created derived from [UserNameSecurityTokenAuthenticator](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Selectors/UserNameSecurityTokenAuthenticator.cs) that has an abstract method as mentioned earlier which contains the logic to validate the digest.

I think this is everything that's needed for the implementation. On top of that, a unit test would be needed. Without client support for this mode, you would need to use HttpClient with a "hand crafted" message (or WCF Client with a message inspector to add it). There's a few details needed to get correct to be secure, and it's these pieces which add the complexity. The nonce and created elements are technically optional, so there's various security aspects to consider if the implementation has them optional too. One option is to keep it simple and use the [spec recommendations](https://docs.oasis-open.org/wss/v1.1/wss-v1.1-spec-pr-UsernameTokenProfile-01.htm#_Toc104276211) which are:

1. It is RECOMMENDED that web service producers reject any UsernameToken not using both nonce and creation timestamps.
2. It is RECOMMENDED that web service producers provide a timestamp “freshness” limitation, and that any UsernameToken with “stale” timestamps be rejected.  As a guideline, a value of five minutes can be used as a minimum to detect, and thus reject, replays.
3. It is RECOMMENDED that used nonces be cached for a period at least as long as the timestamp freshness limitation period, above, and that UsernameToken with nonces that have already been used (and are thus in the cache) be rejected.

If the recommendations work for you, then there's minimum need for providing properties to adjust this behavior and the replay prevention code becomes a lot simpler. If someone else needs to adjust the behavior, then they could do the additional work to support configurability. We have an implementation in CoreWCF that can be used for an expiring nonce cache used to detect replay attacks for Message security already.  

If this is something you want to contribute, I'm happy to have an online meeting to talk about it and be available to help with any challenges.

**Reply by BelvedereHenrique** [reply]
Hello Matt, thanks for your quick and in-depth reply.

For now, I might try to do it manually since the deadline to finish this integration is getting closer, but I'm definitely interested in contributing in my free hours.

An online meeting would be great, please let me know how to proceed.

**Reply by mconnew** [reply]
You might have difficulty doing it manually as you can't bypass handling the token otherwise the security signature won't get checked. The easiest approach would be to tell CoreWCF to use IdenityModel for the token auth (there's a property on ServiceCredentials), and provide a custom SecurityTokenHandler so the security token is still checked which maintains the signature, but at that point you've implemented about 70% of the full solution anyway.  

If you do work out how to do it manually, make sure you implement a basic nonce cache to prevent replay. A very basic implementation would be to have 2 instances of HashSet, one current, one previous. When a message comes in, check both sets, and if not found, add to current. Every 5 minutes (or half of what you want the timeout to be), set previous to current and set current to a new instance of HashSet. You don't need fine grained replay cache expiry unless you are handling tens of thousands of requests per minute.  

To arrange a call, it's probably easier to get on the CoreWCF chat. Our docs still refer to gitter.im, but that's become matrix using the element client. Info [is here](https://github.com/CoreWCF/CoreWCF/blob/main/CONTRIBUTING.md#other-discussions). I just checked and the link still works to open it in a browser.

**Reply by BelvedereHenrique** [reply]
Regarding the cache, this should be done in the server, right? I do not own the server code, I'm simply trying to connect to and use this SOAP WS, as a client.

What I did today was implementing a custom message inspector. In it, I added a custom MessageHeader which includes all the headers properties WS Security asks ( UsernameToken, Username, Password + nonce and CreatedAt)

I plan to create/manipulate these headers every request.
(I might be unawarely breaking tons of security rules here)

I didn't manage to fully test it, but I'm afraid this message headers might get scraped off when transmitted. As I said, I'm not really sure how it works in the background.

Will check out the link, thanks.

**Reply by mconnew** [reply]
I was talking about server things. For client stuff this is the wrong repo, you need dotnet/wcf. Adding it using a message inspector isn't going to work unless you reimplement the while security header implementation as the signature that WCF calculated will be missing/invalid. Adding this feature on the client side should be a lot smaller task if you want to implement it over there. If you want to do it out of box, there's a lot of boiler plate code you'll need to implement, but it shouldn't be too horrendous. It's about 5 classes you would need to implement.

**Reply by BelvedereHenrique** [reply]
Indeed this is the wrong repository for this discussion. I didn't see any discussion tab on the dotnet/wcf repo, that's why I decided to open it here. Let me know if we should stop here and transfer this to an issue on dotnet/wcf.

Regarding the headers, yes, I had to do it manually. I did it using the message inspector BeforeSend method. This WS I'm integrating requires the WS-Security headers in a way that, AFAIK, it breaks conventions sometimes (even text indentation outside XML tags have to be exactly what the WS expects).

Now, I see these headers are being removed/modified during my calls, since the third party WebService is denying my requests because there's no Security header, or is malformed (because something modified it and it is different from what the WS expects it to be). I believe this goes along what you just said. I think  it is related to the binding used.

When I use WSHttpBinding, I see that the WS complains about the prefixes on the security header (therefore, the header is there but it was modified by WCF).
When I use BasicHttpsBinding, there's no security header, even if I manually added it on BeforeSend.

That means WCF is definitely overwriting it somewhere.

Correct if I'm wrong, but I guess I can overwrite this behavior by creating a custom message encoder and use it as a binding in a CustomBinding, is that correct? Do you know what direction should I take, more or less, in this case?

Thanks.

**Reply by BelvedereHenrique** [reply]
@mconnew 
I was investigating it a bit further and I found this issue on WCF repository:
https://github.com/dotnet/wcf/issues/4570

There, you suggested this blog post:
https://justsimplycode.com/2019/11/02/disable-header-100-continue-in-net-core-wcf-client/

I think this could be helpful in our case, since I need to build a fully custom Security XML tag, maybe I can just overwrite whatever is being sent over the wire on this HTTP interceptor.
I already managed to create the XML tag exactly the way the server needs it to be, so maybe I can just transfer the code to this interceptor . What do you think? Is there any other way to do this? A proper way, maybe?

As a reference, this is the header I had to create manually:
```xml
<soapenv:Envelope xmlns:soapenv=http://schemas.xmlsoap.org/soap/envelope/>
    <soapenv:Header>
        <wsse:Security xmlns:soapenv=http://schemas.xmlsoap.org/soap/envelope/
                       xmlns:wsse=http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd
                       xmlns:wsu=http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd>
            <wsse:UsernameToken wsu:Id="{AppName}">
                <wsse:Username>{Username}</wsse:Username>
                <wsse:Password Type=""wsse:PasswordDigest"">{PasswordDigest}</wsse:Password>
                <wsse:Nonce EncodingType=http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#Base64Binary>{Nonce}</wsse:Nonce>
                <wsu:Created>{Created}</wsu:Created>
            </wsse:UsernameToken>
            <ds:Signature xmlns:ds=http://www.w3.org/2000/09/xmldsig#>
                <ds:SignedInfo>
                    <ds:CanonicalizationMethod Algorithm=http://www.w3.org/2001/10/xml-exc-c14n#/>
                    <ds:SignatureMethod Algorithm=http://www.w3.org/2000/09/xmldsig#rsa-sha1/>
                    <ds:Reference URI="{AppName}">
                        <ds:DigestMethod Algorithm=http://www.w3.org/2000/09/xmldsig#sha1/>
                        <ds:DigestValue>{Digest}</ds:DigestValue>
                    </ds:Reference>
                </ds:SignedInfo>
                <ds:SignatureValue>{Signature}</ds:SignatureValue>
            </ds:Signature>
        </wsse:Security>
    </soapenv:Header>
```

I don't own the implementation of the server and If one little space is added or removed, or if a namespace is re-ordered or if a prefix is changed, the request is immediately denied by the WS server.
